sdk/python/tests/compiler/testdata/volume_snapshotop_rokurl.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import kfp.dsl as dsl
+
+
+@dsl.pipeline(
+    name="VolumeSnapshotOp RokURL",
+    description="The fifth example of the design doc."
+)
+def volume_snapshotop_rokurl(rok_url):
+    vop1 = dsl.VolumeOp(
+        name="create_volume_1",
+        resource_name="vol1",
+        size="1Gi",
+        annotations={"rok/origin": rok_url},
+        modes=dsl.VOLUME_MODE_RWM
+    )
+
+    step1 = dsl.ContainerOp(
+        name="step1_concat",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["cat /data/file*| gzip -c >/data/full.gz"],
+        volumes={"/data": vop1.volume}
+    )
+
+    step1_snap = dsl.VolumeSnapshotOp(
+        name="create_snapshot_1",
+        resource_name="snap1",
+        volume=step1.volume
+    )
+
+    vop2 = dsl.VolumeOp(
+        name="create_volume_2",
+        resource_name="vol2",
+        data_source=step1_snap.snapshot,
+        size=step1_snap.outputs["size"]
+    )
+
+    step2 = dsl.ContainerOp(
+        name="step2_gunzip",
+        image="library/bash:4.4.23",
+        command=["gunzip", "-k", "/data/full.gz"],
+        volumes={"/data": vop2.volume}
```

Comment:
It seems that these samples won't work with loops/recursion or whenever a step is reused multiple times. The same volume will be reused and the data overridden. 

Here are some examples/documentations for loop/recursion on which to validate the volume implementation: 

https://www.kubeflow.org/docs/pipelines/sdk/dsl-recursion/
https://github.com/kubeflow/pipelines/blob/master/samples/basic/recursion.py
https://github.com/kubeflow/pipelines/blob/master/sdk/python/tests/compiler/testdata/recursive_while.py

Reply:
@vicaire 
> It seems that these samples won't work with loops/recursion or whenever a step is reused 
multiple times. The same volume will be reused and the data overridden.

Actually, what is going to happen depends on the user. If the user wishes to use a single volume throughout the loop, for all of its iterations, then they should create it before the loop. Similarly, if they have *explicitly* placed a ResourceOp inside a loop, then they expect multiple resources to be created, one for each iteration, so, they need to set the name of the resource accordingly, e.g., based on a PipelineParam value. This will ensure that each loop iteration creates a differently-named resource. Our mechanism supports all of this, it's up to the user to decide their policy.

Here is another example: If the resource has meaning specific to an iteration, e.g. for debugging, then the user should specify a  `PipelineParam` as part of the `resource_name` argument. 
Similarly, the user can choose what volume to mount on a Container, based on the value of `flip_result` in your `recursive_while.py` example.

Another way is for the user to provide a `k8s_resource` directly, and fill the `metadata.generateName`, to allow K8s to determine the final name. No matter what they choose, they will be able to use the final name as an output `PipelineParam` of this `ResourceOp`.

Reply:
Thank you @vkoukis. The current thinking was that we store data in a directory as follows: 

/<subdirectory>/<argo-workflow-UUID>/<POD-UUID>/...

This avoids collision if the same step is executed multiple times (loop, recursion, reuse of a template). 

To make it easy to get the <argo-workflow-UUID> and the <POD-UUID>, we use special Argo strings that get substituted by Argo as the workflow executes: {{{workflow.uid}}}, {{{pod.name}}}

https://github.com/argoproj/argo/blob/master/docs/variables.md

Would it be possible for the DSL to automatically use these strings substituted by Argo to avoid the possibility of a collision?

Reply:
I'm not sure I follow here.
Accessing the volume happens at the filesystem layer.
The task that runs inside the ContainerOp can [and should!] be able to write anywhere in the volume.

Exactly how to keep the results of different executions separate, is actually user-defined policy.
Some users may be creating a new volume for each run, this is actually what we suggest our users do, because it is super easy and efficient to create new volumes/clones. Others, may choose to work under a specific subdirectory, in which case they can always refer to `{{workflow.uid}}` when accessing the volume. A subpath mount may come useful in this case.

Essentially, where the data lives in the volume is user-defined. And they can always use `PipelineParam`s to pass extra information e.g., on the specific directories where they have placed results, from step to step.

This is also important so we can *seed* a pipeline from notebook data. It doesn't make sense to impose a specific `/{{workflow.uid}}` prefix for every single access, since the data may also have been created outside a pipeline context, from within a notebook.